### PR TITLE
Rework Neptune logging

### PR DESCRIPTION
**Pull request by piercus**
* Purely standalone callback (in `train_eraser.py`)
* Include Grid vizualisation
* Only works with Neptune

Grid visualization is organized like :
val:image_I-211117_O07035_T28_F0050

### CODE REVIEW

File: contrib/eraser/training/train_eraser.py [hunk 1]
```diff
@@ -42,9 +48,71 @@
 from lbm.models.unets import DiffusersUNet2DCondWrapper
 from lbm.models.vae import AutoencoderKLDiffusers, AutoencoderKLDiffusersConfig
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py:44]: Let's entirely get rid of it (remove) now. Any reason not to do it?

File: contrib/eraser/training/train_eraser.py [hunk 2]
```diff
@@ -42,9 +48,71 @@
...
 from lbm.trainer import TrainingConfig, TrainingPipeline
-from lbm.trainer.loggers import NeptuneLogger
 from lbm.trainer.utils import StateDictAdapter
 
+class TorchMetricsValidationLogger(Callback):
+    """
+    Logger for logging/accumulating metrics and log them to Neptune.
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]: Re: Neptune: so far it's still agnostic to the underlying logger. I guess it will become Neptune specific in the next PR (visualization part)?

File: contrib/eraser/training/train_eraser.py [hunk 3]
```diff
@@ -42,9 +48,71 @@
...
+
+    def on_validation_batch_end(
+        self,
+        trainer: Trainer,
+        pl_module: TrainingPipeline,
+        outputs: Dict[str, Any],
+        batch: Any,
+        batch_idx: int,
+    ) -> None:
+        # Cross batch/gpu accumulation is handled by pytorch lightning internally.
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]: nit: this comment also apply to `self.log("train/loss", outputs["loss"])` above, right? Same for `self.log(f"val/{key}, ...)` below.

Either repeat or put it at the docstring level (as a note).

File: contrib/eraser/training/train_eraser.py [hunk 4]
```diff
@@ -530,6 +588,7 @@ def main(
     )
 
     trainer = Trainer(
+        log_every_n_steps=10,
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]: Just to make sure: you want this to be enforced? (i.e. this is not a debug left over)

File: contrib/eraser/training/train_eraser.py [hunk 5]
```diff
@@ -42,9 +48,71 @@
 from lbm.models.unets import DiffusersUNet2DCondWrapper
 from lbm.models.vae import AutoencoderKLDiffusers, AutoencoderKLDiffusersConfig
 from lbm.trainer import TrainingConfig, TrainingPipeline
-from lbm.trainer.loggers import NeptuneLogger
 from lbm.trainer.utils import StateDictAdapter
 
+class TorchMetricsValidationLogger(Callback):
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]: naming: right now I'd just rename into `EraserLogger` (or `EraserNeptuneLogger` maybe)

**Review by deltheil — COMMENTED** [review]
PTAL

File: contrib/eraser/training/train_eraser.py (hunk 1 above)
Comment (deltheil) [reply contrib/eraser/training/train_eraser.py:44]: I meant `from lbm.trainer.loggers import NeptuneLogger`, i.e. fully deprecate the former `NeptuneLogger`

File: contrib/eraser/training/train_eraser.py (hunk 4 above)
Comment (piercus) [reply contrib/eraser/training/train_eraser.py]: I clearly want to expose this `log_every_n_steps` arg here (cause it's silently `=50` and it makes things harder to debug)

I may increase this value later, but 10 makes sense (it means we log the loss every 10 optimizer steps)

File: contrib/eraser/training/train_eraser.py (hunk 1 above)
Comment (piercus) [reply contrib/eraser/training/train_eraser.py:44]: `NeptuneLogger` has been removed from `loggers.py`

File: contrib/eraser/training/train_eraser.py (hunk 3 above)
Comment (piercus) [reply contrib/eraser/training/train_eraser.py]: In the docstring

**Comment by piercus** [comment]
Added the multi-gpu support (tested) + grid visualization @deltheil PTAL

File: contrib/eraser/training/train_eraser.py [hunk 6]
```diff
@@ -44,7 +54,245 @@
...
+        self, 
+        num_steps: list[int], 
+        device: torch.device | None = None, 
+    ):
+        super().__init__()
+        self.num_steps = num_steps
+        self.device = device  # can be None; we’ll set it dynamically
+        self.metrics = None  # delay initialization
+
+    def setup(self, trainer: Trainer, pl_module: TrainingPipeline, stage=None):
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]: Since it assumes Neptune only, how about adding an assert like: `assert isinstance(trainer.logger, loggers.NeptuneLogger)`

File: contrib/eraser/training/train_eraser.py [hunk 7]
```diff
@@ -44,7 +54,245 @@
...
 from lbm.trainer import TrainingConfig, TrainingPipeline
 from lbm.trainer.utils import StateDictAdapter
+from dataclasses import field
+from neptune.types import File
+
+class EraserLogger(Callback):
+    """
+    Eraser Logger Callback for Neptune.
+
+    Log LPIPS, DISTS, PSNR.
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]:
```suggestion
    Log LPIPS, DISTS, PSNR metrics, in addition to train and val losses.
```

File: contrib/eraser/training/train_eraser.py [hunk 8]
```diff
@@ -44,7 +54,245 @@
...
+
+    Use lightning self.log function to log metrics, so
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]:
```suggestion
    It uses Lightning's `self.log` method to log metrics, so 
```

File: contrib/eraser/training/train_eraser.py [hunk 9]
```diff
@@ -44,7 +54,245 @@
...
+    """
+    def __init__(
+        self, 
+        num_steps: list[int], 
+        device: torch.device | None = None, 
+    ):
+        super().__init__()
+        self.num_steps = num_steps
+        self.device = device  # can be None; we’ll set it dynamically
+        self.metrics = None  # delay initialization
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py:79]: nit: why don't you do the same for `self.device` (and remove it from the CTOR args)?

File: contrib/eraser/training/train_eraser.py (hunk 6 above)
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]:
```suggestion
    def setup(self, trainer: Trainer, pl_module: TrainingPipeline, stage=None) -> None:
```

File: contrib/eraser/training/train_eraser.py [hunk 10]
```diff
@@ -44,7 +54,245 @@
 from lbm.models.vae import AutoencoderKLDiffusers, AutoencoderKLDiffusersConfig
 from lbm.trainer import TrainingConfig, TrainingPipeline
 from lbm.trainer.utils import StateDictAdapter
+from dataclasses import field
+from neptune.types import File
+
+class EraserLogger(Callback):
+    """
+    Eraser Logger Callback for Neptune.
```
Comment (deltheil) [inline contrib/eraser/training/train_eraser.py]:
```suggestion
    Eraser Logger Callback made for Neptune.
```

**Review by deltheil — COMMENTED** [review]
Some final nit-s, PTAL

**Review by deltheil — APPROVED** [review]
